Move _handle_aliases() out of basic.py

samdoran opened the pull request:
##### SUMMARY

Part of the work to enable controller side arg spec parsing/validation.

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Feature Pull Request

##### COMPONENT NAME
`lib/ansible/module_utils/basic.py`
`lib/ansible/module_utils/common/parameters.py`

##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes -->
```paste below
2.8
```

samdoran committed:
Move _handle_aliases out of basic.py

abadger commented on lib/ansible/module_utils/basic.py:
```diff
@@ -1630,32 +1614,7 @@ def _check_locale(self):
-                    param[k] = param[alias]
-
-        return aliases_results
+        return handle_aliases(self.params, self.argument_spec, self._legal_inputs, spec, param)
```
This is a private function (starts with an underscore) so I'd hope we could move this without having to leave a stub for backwards compatibility.  If you want to remove the stub in a separate commit/PR just in case, though, I can live with that.

abadger replied:
After taking a look at the new function, I could see leaving the stub function in so that it can establish default values for arg_spec and params.

abadger commented on lib/ansible/module_utils/common/parameters.py:
```diff
@@ -0,0 +1,68 @@
+SEQUENCETYPE = frozenset, KeysView, Sequence
+
+
+def handle_aliases(params, argument_spec, legal_inputs=None, spec=None, param=None):
```
We don't need both params and param.  Same with argument_spec and spec.

abadger commented on lib/ansible/module_utils/common/parameters.py:
```diff
@@ -0,0 +1,68 @@
+
+def handle_aliases(params, argument_spec, legal_inputs=None, spec=None, param=None):
+    if legal_inputs is None:
+        legal_inputs = ['_ansible_%s' % k for k in PASS_VARS]
```
We should probably make this a python module-level constant.

abadger replied:
That way we don't have to set it every time we call this function.

abadger replied:
Hm.... I just noticed that legal_inputs is being modified below as well.  That's not great practice unless we know that we could be dealing with really large lists.  That's not the case here so I think we should do something like:
``` python
# Module global scope
DEFAULT_LEGAL_PARAMS = ['_ansible_%s' % k for k in PASS_VARS]
[...]
def handle_aliases([...]):
    if legal_inputs is None:
        legal_inputs = DEFAULT_LEGAL_PARAMS[:]
```

abadger commented on lib/ansible/module_utils/common/parameters.py:
```diff
@@ -0,0 +1,68 @@
+def handle_aliases(params, argument_spec, legal_inputs=None, spec=None, param=None):
+    if legal_inputs is None:
+        legal_inputs = ['_ansible_%s' % k for k in PASS_VARS]
+    # this uses exceptions as it happens before we can safely call fail_json
```
We can remove this comment as nearly all of the code we extract should be raising exceptions.

abadger commented on lib/ansible/module_utils/basic.py:
```diff
@@ -1630,32 +1614,7 @@ def _check_locale(self):
                            to_native(e), exception=traceback.format_exc())
 
     def _handle_aliases(self, spec=None, param=None):
-        # this uses exceptions as it happens before we can safely call fail_json
```
This comment might need to stay in the stub function.  At least until we move fail_json() and exit_json() out.

abadger commented on lib/ansible/module_utils/common/parameters.py:
```diff
@@ -0,0 +1,68 @@
+        if aliases is None:
+            continue
+        if not isinstance(aliases, SEQUENCETYPE) or isinstance(aliases, (binary_type, text_type)):
+            raise Exception('internal error: aliases must be a list or tuple')
```
I am really bad at choosing exceptions but we probably should take this opportunity to use a less generic exception.  Maybe ValueError on line 57 and TypeError on line 61.

abadger commented on lib/ansible/module_utils/common/parameters.py:
```diff
@@ -0,0 +1,68 @@
+            if alias in param:
+                param[k] = param[alias]
+
+    return aliases_results
```
legal_inputs was also being mutated above.  So we want to return legal_inputs in addition to aliases_results.

samdoran committed:
Refinements

- return legal_inputs and update class properties
- remove redundant arguments from method and handle in caller
- add better exception types to method

abadger commented on lib/ansible/module_utils/common/parameters.py:
```diff
@@ -0,0 +1,64 @@
+
+def handle_aliases(argument_spec, params, legal_inputs=None):
+    if legal_inputs is None:
+        legal_inputs = DEFAULT_LEGAL_PARAMS[:]
```
Thinking about this more, we should probably make a copy if legal_inputs existed too.

samdoran replied:
I believe I did this correctly:

```python
# snip
    if legal_inputs is None:
        legal_inputs = DEFAULT_LEGAL_PARAMS[:]
    elif legal_inputs:
        legal_inputs = legal_inputs[:]
```

mattclay commented:
CI failure in unit tests: https://app.shippable.com/github/ansible/ansible/runs/93920/8/tests

samdoran committed:
Fix unit tests

abadger commented on lib/ansible/module_utils/common/parameters.py:
```diff
@@ -0,0 +1,64 @@
+            raise ValueError("internal error: required and default are mutually exclusive for %s" % k)
+        if aliases is None:
+            continue
+        if not isinstance(aliases, SEQUENCETYPE) or isinstance(aliases, (binary_type, text_type)):
```
Instead of moving SEQUENCETYPE, I think we can switch this to using is_iterable.

``` python
from ansible.module_utils.common.collections import is_iterable
[...]
[Line 56 becomes]
if not is_iterable(aliases):
```

samdoran committed:
Address review comments

abadger commented on lib/ansible/module_utils/common/parameters.py:
```diff
@@ -0,0 +1,121 @@
+    return no_log_values
+
+
+def list_deprications(argument_spec, params):
```
spelling: list_deprecations

samdoran committed:
Fix tests and add some docstrings

samdoran commented:
Rebased to fix the sanity test errors.

mattclay commented:
@samdoran Rebasing is only needed if there are merge conflicts. Otherwise a re-run of the failed jobs on Shippable will suffice, since each run automatically rebases before running tests.

mattclay commented:
Although in this case, since it's marked `stale_ci`, a full run makes sense.

samdoran commented:
@mattclay I figured I needed to rebase due to the module doc type arguments being tightened up. But now that I think about it more, you're correct that I didn't need to rebase. Thanks for the info!

samdoran commented:
I think the remaining test failures are unrelated to this PR. I will want to rebase this once #48628 is merged.

sivel commented on lib/ansible/module_utils/common/parameters.py:
```diff
@@ -0,0 +1,124 @@
+    'version': 'ansible_version',
+}
+
+DEFAULT_LEGAL_PARAMS = ['_ansible_%s' % k for k in PASS_VARS]
```
With this PR, we end up doing this same work 2 times.  Once here and once in `AnsibleModule.__init__`.  I think we should likely stop doing it in `AnsibleModule.__init__` and just use this value.

sivel commented on lib/ansible/module_utils/common/parameters.py:
```diff
@@ -0,0 +1,124 @@
+    a list of legal inputs, or None."""
+
+    if legal_inputs is None:
+        legal_inputs = DEFAULT_LEGAL_PARAMS[:]
```
Looking at this, I wonder if maybe we should just do this here, and start out `self._legal_inputs` from `AnsibleModule.__init__` as an empty list, and then there would be no reason to pass in `legal_inputs` to begin with.  Just make this function responsible for setting it, since that is the end result anyway.

sivel commented on lib/ansible/module_utils/common/parameters.py:
```diff
@@ -0,0 +1,124 @@
+    return aliases_results, legal_inputs
+
+
+def _return_datastructure_name(obj):
```
It appears that everything from here down, should not be in this PR.  As it seems that https://github.com/ansible/ansible/pull/48628/files is meant to handle these.

samdoran committed:
Handle legal inputs inside handle_aliases

samdoran committed:
Add unit tests for handle_aliases

sivel commented on lib/ansible/module_utils/basic.py:
```diff
@@ -1583,32 +1569,15 @@ def _check_locale(self):
-        return aliases_results
+        # this uses exceptions as it happens before we can safely call fail_json
+        alias_results, legal_inputs = handle_aliases(spec, param)
+        self._legal_inputs = legal_inputs
```
Technically, you could remove this line, and just use:

```
alias_results, self._legal_inputs = handle_aliases(spec, param)
```

samdoran replied:
Much nicer.

sivel commented on test/units/module_utils/common/parameters/test_handle_aliases.py:
```diff
@@ -0,0 +1,95 @@
+        DEFAULT_LEGAL_INPUTS + ['name'],
+    )
+
+    assert result == handle_aliases(argument_spec, params)
```
It looks like you will need to sort the legal inputs before comparing.

Something like:

```
diff --git a/test/units/module_utils/common/parameters/test_handle_aliases.py b/test/units/module_utils/common/parameters/test_handle_aliases.py
index 0db4c2ee7a..2b1e13cd95 100644
--- a/test/units/module_utils/common/parameters/test_handle_aliases.py
+++ b/test/units/module_utils/common/parameters/test_handle_aliases.py
@@ -59,12 +59,15 @@ def test_handle_aliases_basic():
         'nick': 'foo',
     }
 
-    result = (
+    expected = (
         {'surname': 'name', 'nick': 'name'},
         DEFAULT_LEGAL_INPUTS + ['name', 'surname', 'nick'],
     )
 
-    assert result == handle_aliases(argument_spec, params)
+    expected[1].sort()
+    result = handle_aliases(argument_spec, params)
+    result[1].sort()
+    assert expected == result
 
 
 def test_handle_aliases_value_error():
```

That diff is for the next testa actually.

samdoran committed:
Sort legal inputs in unit tests
